### ISSUE TITLE
Inserting multiple breakpoints in one function throws an error

### ISSUE THREAD

**Issue by ateucher** [rstudio/rstudio#8451]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.4.1055
    OS Version      : MacOS 10.14.6
    R Version       : 4.0.2

### Steps to reproduce the problem

1. Ensure the R check length 1 condition is set to to `true` or (better) `verbose`:

```r
Sys.setenv("_R_CHECK_LENGTH_1_LOGIC2_"="verbose")
```

2. Write a function definition in a file. eg:

```r
fun <- function(x) {
  y <- mean(x)
  y
}
```

3. Save and source the file (via the "Source" button or `source("myfile.R")`)

4. Click in the gutter on two different lines inside the function body to set two breakpoints

5. An error and failure report is immediately printed to the console and RStudio hangs for a short time and then R aborts:

```
> debugSource('myfile.R', echo=TRUE)
 ----------- FAILURE REPORT -------------- 
 --- failure: length > 1 in coercion to logical ---
 --- srcref --- 
: 
 --- package (from environment) --- 
package:stats
 --- call from context --- 
.rs.setFunctionBreakpoints("fun", env, lapply(steps, function(step) {
    step$at
}))
 --- call from argument --- 
length(steps) == 0 || nchar(steps) == 0
 --- R stacktrace ---
where 1: .rs.setFunctionBreakpoints("fun", env, lapply(steps, function(step) {
    step$at
}))
where 2: withCallingHandlers(expr, warning = function(w) if (inherits(w, 
    classes)) tryInvokeRestart("muffleWarning"))
where 3: suppressWarnings(.rs.setFunctionBreakpoints("fun", env, lapply(steps, 
    function(step) {
        step$at
    })))
where 4: (function (fileName, encoding, breaklines, local) 
{
    envir <- if (isTRUE(local)) {
        sys.frames()[[1]]
    }
    else if (identical(local, FALSE)) {
        .GlobalEnv
    }
    else if (is.environment(local)) {
        local
    }
    else {
        stop("'local' must be TRUE, FALSE or an environment")
    }
    env <- new.env(parent = emptyenv())
    env$fun <- .rs.makeSourceEquivFunction(fileName, encoding, 
        envir)
    breakSteps <- character()
    if (length(breaklines) > 0) {
        steps <- .rs.getFunctionSteps(env$fun, "fun", breaklines)
        breakSteps <- unlist(lapply(steps, function(step) step$at))
        suppressWarnings(.rs.setFunctionBreakpoints("fun", env, 
            lapply(steps, function(step) {
                step$at
            })))
    }
    env$fun()
    env$fun <- .rs.removeBreakpoints(env$fun)
    breakSteps <- breakSteps[nchar(breakSteps) > 6]
    for (steps in breakSteps) {
        step <- as.numeric(strsplit(breakSteps, ",")[[1]][3])
        op <- deparse(body(env$fun)[[2]][[2]][[step]][[1]])
        if (op == "<-" || op == "=") 
            eval(body(env$fun)[[2]][[2]][[step]], envir = globalenv())
    }
    return(NULL)
})("myfile.R", "unknown", 2:3, FALSE)
where 5: .Call("rs_debugSourceFile", enc2utf8(fileName), encoding, local)
where 6: debugSource("myfile.R", echo = TRUE)

 --- value of length: 2 type: logical ---
[1] FALSE FALSE
 --- function from context --- 
function (functionName, envir, steps) 
{
    functionName <- .rs.unquote(functionName)
    if (length(steps) == 0 || nchar(steps) == 0) {
        fun <- get(functionName, envir = envir)
        if (.rs.isTraced(fun)) {
            suppressMessages(untrace(what = functionName, where = envir))
        }
    }
    else {
        suppressMessages(trace(what = functionName, where = envir, 
            at = lapply(strsplit(as.character(steps), ","), as.numeric), 
            tracer = browser, print = FALSE))
        lockedBinding <- FALSE
        tryCatch({
            if (bindingIsLocked(functionName, envir)) {
                unlockBinding(functionName, envir)
                lockedBinding <- TRUE
            }
            body(envir[[functionName]]@.Data) <- .rs.tracedSourceRefs(body(envir[[functionName]]@.Data), 
                body(envir[[functionName]]@original))
        }, finally = {
            if (lockedBinding) {
                lockBinding(functionName, envir)
            }
        })
    }
    return(functionName)
}
<environment: 0x7fe601d0d908>
 --- function search by body ---
 ----------- END OF FAILURE REPORT -------------- 
Error in length(steps) == 0 || nchar(steps) == 0 : 
  'length(x) = 2 > 1' in coercion to 'logical(1)'
```
If I set `_R_CHECK_LENGTH_1_LOGIC2_=true`, only the error message is printed (not the verbose failure report) and R doesn't crash.

### Describe the problem in detail

If I try to set two breakpoints in a function when `_R_CHECK_LENGTH_1_LOGIC2_=verbose` or `_R_CHECK_LENGTH_1_LOGIC2_=true`, an error occurs.

### Describe the behavior you expected

I should be able to set two breakpoints in a function. If I set `_R_CHECK_LENGTH_1_LOGIC2_=false`, both breakpoints are activated as expected. I understand this is not a common thing to do - I discovered the bug when I accidentally clicked twice inside a function and the error occurred.

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting a RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ronblum** [comment]
@ateucher Thank you for bringing this up! We'll review this issue as we continue development of RStudio.

**Comment by cderv** [comment]
Hi @ronblum, 

This is still an issue I encounter often. Would it be possible to reopen it ? Otherwise, I can open a new one but it seems that this one as a good example. 

It would be great to fix this bug at some point.Thanks!

**Comment by mikebessuille** [comment]
@cderv does this happen in the command line using just R?

**Comment by cderv** [comment]
@mikebessuille I'll need to try that. I do it all the time within the IDE because breakpoint is for me an IDE feature. 
So that I can try, What would be the equivalent of setting breakpoint (the red dot in the IDE source) with command line and just R ?

**Comment by mikebessuille** [comment]
I believe you can use the browser() function but I am not an R expert...

**Comment by jmcphers** [comment]
This is an error in `setFunctionBreakpoints` which is part of RStudio's code, so it is a bug in our debugger.

**Comment by jmcphers** [comment]
R _does_ have similar functionality, see `setBreakpoint()`: https://www.rdocumentation.org/packages/utils/versions/3.6.2/topics/findLineNum

**Comment by miserman** [comment]
This should just need a change to the condition in https://github.com/rstudio/rstudio/blob/67bce36c3d3293080e9f1a174749111da4f58fe6/src/cpp/session/modules/SessionBreakpoints.R#L275 such as to one of
```r
length(steps) == 0 || (length(steps) == 1 && nchar(steps) == 0)
length(steps) == 0 || nchar(steps[[1]]) == 0
length(steps) == 0 || any(nchar(steps) == 0)
```
depending on the intention.

This is a particularly annoying issue after R 4.2.0.

**Comment by sanjmeh** [comment]
I have been consistently encountering the following warning messages while using the Rstudio debugger, and after investigating the issue, I discovered that others are facing the same problem:

```
Warning messages:
1: In length(steps) == 0 || nchar(steps) == 0 :
  'length(x) = 2 > 1' in coercion to 'logical(1)'
2: In length(steps) == 0 || nchar(steps) == 0 :
  'length(x) = 2 > 1' in coercion to 'logical(1)'

```

Initially, I was unable to correlate these warnings with the debugger, and I spent considerable time searching my code for the offending lines `nchar(steps) == 0` and `length(steps) == 0`, but to no avail.

If the warning messages are harmless, could we at least suppress them to prevent confusion for users?

I strongly urge the maintainers to address this issue, as I can guarantee that thousands of R programmers are affected by it, even if they are not all raising a GitHub issue. This bug not only causes confusion but also wastes valuable time for users who are trying to identify the problem in their own code. Fixing this issue would greatly improve the user experience for many R programmers.
